api/dbadapters/bigquery.ts

```diff
@@ -199,13 +200,68 @@ export class BigQueryDbAdapter implements IDbAdapter {
       `SELECT * FROM ${CACHED_STATE_TABLE_NAME}`,
       5000 // TODO: Add pagination for 5000+ rows
     );
-    const persistedMetadata = rows.map(row => {
-      const encodedProto = Buffer.from(row.proto, "base64");
+    const persistedMetadata = rows.map((row: IMetadataRow) => {
+      const encodedProto = Buffer.from(row.metadata_proto, "base64");
       return dataform.PersistedTableMetadata.decode(encodedProto);
     });
     return persistedMetadata;
   }
 
+  public async persistStateMetadata(executionGraph: dataform.IExecutionGraph): Promise<void> {
+    const { actions } = executionGraph;
+    const metadataTableCreateQuery = `
+      CREATE TABLE IF NOT EXISTS \`${CACHED_STATE_TABLE_NAME}\` (
+        target_name STRING,
+        metadata_json STRING,
+        metadata_proto STRING
+      )
+    `;
+    await this.runQuery(metadataTableCreateQuery);
+    const tableMetadataMap = new Map<dataform.ITarget, IBigQueryTableMetadata>();
+    await Promise.all(
+      actions.map(async action => {
+        tableMetadataMap.set(action.target, await this.getMetadata(action.target));
+      })
+    );
+    const updateQueries = actions.map(action => {
+      const definitionHash = hashTableDefinition(action);
```

Comment:
all of this code to do with computing `PersistedTableMetadata` as a result of table metadata feels like it should probably be in a common utility location.